packages/backend/src/open_payments/payment/outgoing/routes.ts

```diff
@@ -69,35 +76,63 @@ async function getOutgoingPayment(
   )
 }
 
-export type CreateBody = {
+type CreateBodyBase = {
   walletAddress: string
-  quoteId: string
   metadata?: Record<string, unknown>
 }
 
+type CreateBodyFromQuote = CreateBodyBase & {
+  quoteId: string
+}
+
+type CreateBodyFromIncomingPayment = CreateBodyBase & {
+  incomingPayment: string
+  debitAmount: Amount
+}
+
+function isCreateFromIncomingPayment(
+  body: CreateBody
+): body is CreateBodyFromIncomingPayment {
+  return 'incomingPayment' in body && 'debitAmount' in body
+}
+
+export type CreateBody = CreateBodyFromQuote | CreateBodyFromIncomingPayment
+
 async function createOutgoingPayment(
   deps: ServiceDependencies,
   ctx: CreateContext<CreateBody>
 ): Promise<void> {
   const { body } = ctx.request
-
-  const quoteUrlParts = body.quoteId.split('/')
-  const quoteId = quoteUrlParts.pop() || quoteUrlParts.pop() // handle trailing slash
-  if (!quoteId) {
-    throw new OpenPaymentsServerRouteError(
-      400,
-      'Invalid quote id trying to create outgoing payment',
-      { requestBody: body }
-    )
-  }
-
-  const outgoingPaymentOrError = await deps.outgoingPaymentService.create({
+  let options: OutgoingPaymentCreateBaseOptions = {
     walletAddressId: ctx.walletAddress.id,
-    quoteId,
     metadata: body.metadata,
     client: ctx.client,
     grant: ctx.grant
-  })
+  }
+  if (isCreateFromIncomingPayment(body)) {
+    options = {
+      ...options,
+      incomingPayment: body.incomingPayment,
+      debitAmount: body.debitAmount
+    } as CreateFromIncomingPayment
+  } else {
+    const quoteUrlParts = body.quoteId.split('/')
+    const quoteId = quoteUrlParts.pop() || quoteUrlParts.pop() // handle trailing slash
+    if (!quoteId) {
+      throw new OpenPaymentsServerRouteError(
+        400,
+        'Invalid quote id trying to create outgoing payment'
+      )
+    }
+    options = {
+      ...options,
+      quoteId
+    } as CreateFromQuote
+  }
+
+  const outgoingPaymentOrError = await deps.outgoingPaymentService.create(
+    options as CreateOutgoingPaymentOptions
```

Comment:
Huh, I guess TS isn't smart enough to know that at this point options will be (correctly) either `CreateFromIncomingPayment` or `CreateFromQuote`, which are basically `CreateOutgoingPaymentOptions` ?

Reply:
Yeah this kinda sucks. It stems from the union type and initializing the options and mutating it. I took another stab at it and think this is much improved:  [7ef6492](https://github.com/interledger/rafiki/pull/2654/commits/7ef649269ff722fbf5f8981701cb23ec26cb91ec)

I also considered putting everything under the typeguard (forming options, calling `outgoingPaymentService.create(options)`) but I preferred not to duplicate that stuff in each block and wanted to keep them more focused on what actually depended on the type guard.